### PR TITLE
Bugfix/crossfade filters when activated

### LINKED ISSUE #932: Click when engaging filter from 0
### Please describe the problem:

When turning up a filter (HPF most notably) a click is heard. This  is not good for smooth filter modulations.

### What is the expected behavior?

No click

### Is there a relevant Pull request?

No

### What hardware did you reproduce it with?

OLED

### What firmware did you reproduce it with?

Nightly

### What is the firmware name:

1.1.0-NIGHTLY-SESSEC7

### If possible provide the steps to reproduce the issue and upload additional media:

Turn up the filter while a tone is playing

### PR DESCRIPTION
**Pull request by m-m-adams**
Adds a crossfade when filters are first turned on to avoid clicks and pops

Fixes #932 

In order to do that makes a number of other supporting changes:
1. Refactors filter functions to remove extraneous doFilter booleans, that is now passed by setting the filter off
2. Modifies getFilterDisplayName to not turn the filters back on
3. Adds new functions for toggling filter modes including off - not used in this PR, will be part of a future PR to enable filters in FM

Also marks rendering functions as hot, which adds a couple voices overall

### CODE REVIEW

File: src/deluge/dsp/filter/filter_set.h [hunk 1]
```diff
@@ -59,12 +60,17 @@ class FilterSet {
...
+	// all filters share a state. This is fine since they just hold plain data and initialization is handled by
+	// reset/configure calls.  This is faster than using a variant at the cost of not throwing on unitialized access.
+	// However since there are no invariants to uphold, the worst case scenario is an audio glitch so whatever
+	// This cuts 250 bytes from the size of this class, which is fairly significant since every voice has a filterset
+	char lpfilterstate[lpfilterSize];
+	char hpfilterstate[hpfilterSize];
+	SVFilter& lpsvf = *(SVFilter*)lpfilterstate;
+	LpLadderFilter& lpladder = *(LpLadderFilter*)lpfilterstate;
+	HpLadderFilter& hpladder = *(HpLadderFilter*)hpfilterstate;
+	SVFilter& hpsvf = *(SVFilter*)hpfilterstate;
```
Comment (sapphire-arches) [inline src/deluge/dsp/filter/filter_set.h:73]: Can we use a `union` instead? Same no-throw behavior but without the questionable casting, I think.

File: src/deluge/dsp/filter/filter.h [hunk 2]
```diff
@@ -54,23 +57,68 @@ class Filter {
...
+		if (fade) {
+			dryFade = 1;
+			wetLevel = 0;
+		}
+	}
+
+	inline void updateBlend() {
+		// fades over around 500 samples
+		dryFade = dryFade * 0.99;
+		wetLevel = (q31_t)(ONE_Q31 * (1 - dryFade));
```
Comment (sapphire-arches) [inline src/deluge/dsp/filter/filter.h:120]: does this eventually get all the way to `ONE_Q31`? I imagine so, but do we need an explicit clamp?

File: src/deluge/dsp/filter/filter.h [hunk 3]
```diff
@@ -54,23 +57,68 @@ class Filter {
...
 	 */
-	void filterStereo(q31_t* startSample, q31_t* endSample) {
-		static_cast<T*>(this)->doFilterStereo(startSample, endSample);
-		;
+	[[gnu::hot]] void filterStereo(q31_t* startSample, q31_t* endSample) {
+		if (wetLevel == ONE_Q31) {
+			static_cast<T*>(this)->doFilterStereo(startSample, endSample);
+		}
+		else {
+			memcpy(blendBuffer, startSample, (endSample - startSample) * sizeof(q31_t));
```
Comment (sapphire-arches) [inline src/deluge/dsp/filter/filter.h:91]:
```suggestion
			std::copy(startSample, endSample, &blendBuffer[0]);
```
Double check the binary size but I think it should be fine.

File: src/deluge/dsp/filter/filter_set.h (hunk 1 above)
Comment (m-m-adams) [reply src/deluge/dsp/filter/filter_set.h:73]: Smart - I thought you could only use POD structs in unions but looks like that may not be the case anymore

File: src/deluge/dsp/filter/filter.h (hunk 2 above)
Comment (m-m-adams) [reply src/deluge/dsp/filter/filter.h:120]: yeah it does, it gets there by underflowing after about 500 samples

File: src/deluge/dsp/filter/filter.h (hunk 3 above)
Comment (m-m-adams) [reply src/deluge/dsp/filter/filter.h:91]: we have a faster arm assembly memcpy implementation than libc/c++ and it makes a meaningful difference

It's in c_lib_alternatives.s

File: src/deluge/dsp/filter/filter_set.cpp [hunk 4]
```diff
@@ -198,10 +192,9 @@ int32_t FilterSet::setConfig(int32_t lpfFrequency, int32_t lpfResonance, bool do
...
 void FilterSet::reset() {
-	hpladder.reset();
-	lpsvf.reset();
-	hpsvf.reset();
-	lpladder.reset();
-	noiseLastValue = 0;
+	hpfilter.ladder.reset();
+	lpfilter.svf.reset();
+	hpfilter.svf.reset();
+	lpfilter.ladder.reset();
```
Comment (sapphire-arches) [inline src/deluge/dsp/filter/filter_set.cpp:198]: this looks a little suspect now, since the filters share storage.

Comment (m-m-adams) [reply src/deluge/dsp/filter/filter_set.cpp:198]: oh good point - I don't think it's necessary at all

Comment (m-m-adams) [reply src/deluge/dsp/filter/filter_set.cpp]: It is necessary but zeroing the storage works - the reset for each filter just sets the state to 0 anyway

Comment (sapphire-arches) [reply src/deluge/dsp/filter/filter_set.cpp]: cool, I guess just leave a comment to the effect of "all zeros must represent a reset state" on the union? the rest of this PR looks fine.

Comment (m-m-adams) [reply src/deluge/dsp/filter/filter_set.cpp]: done, thanks

**Pull request merged**